### ISSUE TITLE
PNG color order

### ISSUE THREAD

**Issue by JinShil** [nothings/stb#181]
On my display the JPG decoder produces the correct color order, but the PNG decoder swaps my red and blue. I think this is because the PNG format is big endian and my platform is little endian, but I'm only speculating.  

I've corrected for this using an expensive post processor to reverse the color components, but I'm wondering if perhaps I'm simply using stb_image incorrectly.  Please advise.

**Comment by nothings** [comment]
No, that's not supposed to happen.

If you're using the exact same code and the only thing that changes is the image you're feeding to the code, then please post a pair of images (a JPG that works and a PNG that fails) or email them to me, so I can check whether it's something about the images (vs something about your compiler/platform/etc).

**Comment by JinShil** [comment]
Thanks for the reply.

Here are two sample images for demonstration:

PNG File (right-most color is supposed to be white at 50% intensity - i.e. partially transparent)
![colortest](https://cloud.githubusercontent.com/assets/6023641/10069017/3e472dba-62e4-11e5-9252-7fd50ff240dc.png)

JPG File (right-most color is white since JPGs don't support transparency)
![colortest](https://cloud.githubusercontent.com/assets/6023641/10069021/42059806-62e4-11e5-8e14-4822fc05396a.jpg)

Here's the JPG on my platform (ARM Cortex-M4F):
![wp_20150924_17_54_27_pro](https://cloud.githubusercontent.com/assets/6023641/10069185/6c978150-62e5-11e5-9587-48e253ce8a22.jpg)

And here's the PNG (notice the red and blue are swapped):
![wp_20150924_17_42_57_pro](https://cloud.githubusercontent.com/assets/6023641/10069194/803d1d8c-62e5-11e5-8615-a65407557901.jpg)

I was able to get rid of my expensive post processor by modifying [lines 3112~3114](https://github.com/nothings/stb/blob/master/stb_image.h#L3112) to...
```
out[0] = (stbi_uc)b;
out[1] = (stbi_uc)g;
out[2] = (stbi_uc)r;
```
... and [lines 4377~4379](https://github.com/nothings/stb/blob/master/stb_image.h#L4377) to...
```
palette[i*4+2] = stbi__get8(s);
palette[i*4+1] = stbi__get8(s);
palette[i*4+0] = stbi__get8(s);
```
... (i.e. simply swapping red and blue), but I'm still wondering why I had to do that.  

My calling code is the same for both file formats using `stbi_load_from_callbacks`, so I don't suspect any problem there.

I see the same problem whether or not I read from my board's SDCard, or stream a PNG from the web.

I'm compiling with [GNU Tools for ARM](https://launchpad.net/gcc-arm-embedded).

**Comment by nothings** [comment]
Have you enabled 'stbi_convert_iphone_png_to_rgb'? That does an R/B swap.

Also, the two changes you posted will R/B swap JPGs and paletted-PNGs, but the PNG you posted doesn't have a palette, so the changes you've made seem like they will R/B swap the sample JPG but not the sample PNG.

So, while I could be wrong, I'm guessing you're dealing with a display that expects BGR, and stb_image outputs RGB. So let's keep your display device out of it: what are first four bytes output by stbi_load_... for these two images using an unmodified stb_image.h?

**Comment by JinShil** [comment]
> So, while I could be wrong, I'm guessing you're dealing with a display that expects BGR, and stb_image outputs RGB.

Yep, I think you're right.  As I was trying to figure things out I had also modified the JPG code as I linked above, and forgot about it (even though I pointed it out in the comment above, I thought it was PNG code).  So the reason JPG was working was because I swapped R and B for JPG, too.

In addition to the changes I listed above, I had also changed the code for PNGs between [lines 4438 and 4439](https://github.com/nothings/stb/blob/master/stb_image.h#L4438-L4439) for images without a palette.
```
if (s->img_out_n == 4)
{
	uint8_t* _z = (uint8_t*)z->out;
	for(int i = 0; i < s->img_x * s->img_y * s->img_out_n; i+=s->img_out_n)
	{
		uint8_t _t = _z[i];
		_z[i] = _z[i+2];
		_z[i+2] = _t;
	}
}
```

In summary, I'm tripping over myself as I try to get this to work with my display.  I'm terribly sorry for the distraction. Closing.
